### PR TITLE
Add Windows Active Directory Mixin

### PR DESCRIPTION
**Pull request by BominRahmani**
Adds dashboards and alerts for the Windows Active Directory mixin.

# Notes to reviewers: 
This was done using the new grafonnet library for the first time. There may be some weird things being done, I am more than happy to refactor if this is the case.

The dashboards are available in the cloud environment for review, You must select the datasource from the selectors, since this defaults it to prometheus. Same with the logs.

# Dashboards:
- Windows Active Directory overview
- Windows logs
image
image

# Alerts:
- WindowsActiveDirectoryHighPendingReplicationOperations: There is a high number of pending replication operations in Active Directory. A high number of pending operations sustained over a period of time can indicate a problem with replication.
- WindowsActiveDirectoryHighReplicationSyncRequestFailures: There are a number of replication synchronization request failures. These can cause authentication failures, outdated information being propagated across domain controllers, and potentially data loss or inconsistencies.'
- WindowsActiveDirectoryHighPasswordChange: There is a high number of password changes. This may indicate unauthorized changes or attacks.
- WindowsActiveDirectoryMetricsDown: Windows Active Directory metrics are down.

### CODE REVIEW

File: windows-active-directory-mixin/README.md
Comment (StefanKurek) [inline windows-active-directory-mixin/README.md:11]: I believe these should be the names of the alerts and not the thresholds

File: windows-active-directory-mixin/README.md
Comment (StefanKurek) [inline windows-active-directory-mixin/README.md:64]: again, should have the names of alerts and not thresholds

File: windows-active-directory-mixin/alerts/alerts.libsonnet
Comment (StefanKurek) [inline windows-active-directory-mixin/alerts/alerts.libsonnet:17]: The description for this alert doesn't really match up with reality. The description sounds like we're talking about something like too many bind operations compared to total operations.

Also, does this math make sense? it seems like if there is a single bind operation every minute, (or even every 5 minutes), we're going to end up with 50% which hits the threshhold to trigger the alert?

File: windows-active-directory-mixin/alerts/alerts.libsonnet
Comment (StefanKurek) [inline windows-active-directory-mixin/alerts/alerts.libsonnet:2]: looks like you're missing your high replication issues alert

File: windows-active-directory-mixin/README.md
Comment (StefanKurek) [inline windows-active-directory-mixin/README.md:59]: shouldn't we relabel the job to something like "integration/active_directory" or something like that to tell the difference between different windows jobs (like IIS)?

**Review by StefanKurek — CHANGES_REQUESTED** [review]
A few issues. Haven't looked at the dashboard yet. Waiting until spec approval.

File: windows-active-directory-mixin/README.md
Comment (dpaasman00) [reply windows-active-directory-mixin/README.md:11]: ^^^

File: windows-active-directory-mixin/README.md
Comment (dpaasman00) [inline windows-active-directory-mixin/README.md:27]:
```suggestion
Windows Active Directory system logs are enabled by default in the `config.libsonnet` and can be disabled by setting `enableLokiLogs` to `false`. Then run `make` again to regenerate the dashboard:
```

File: windows-active-directory-mixin/README.md
Comment (dpaasman00) [inline windows-active-directory-mixin/README.md:24]:
```suggestion
The Windows Active Directory logs dashboard provides details on incoming AD logs.
```
"system" doesn't seem right unless they are actual system logs, so maybe "AD" or something else

File: windows-active-directory-mixin/README.md
Comment (dpaasman00) [inline windows-active-directory-mixin/README.md:37]:
```suggestion
For the selectors to properly work with AD logs ingested into your logs datasource, please also include `instance` and `job` labels in the [scrape configs](https://grafana.com/docs/loki/latest/clients/promtail/configuration/#scrape_configs) so they match the labels for ingested metrics.
```

**Review by dpaasman00 — COMMENTED** [review]
Few comments for the moment

File: windows-active-directory-mixin/README.md
Comment (StefanKurek) [reply windows-active-directory-mixin/README.md]: that's interesting. because for IIS i thought we were changing the label to NOT be the generic "integrations/windows" so that we could filter by job and only have the IIS metrics. either way, if vitaly suggested that, then i'm ok with it as well

File: windows-active-directory-mixin/README.md
Comment (StefanKurek) [inline windows-active-directory-mixin/README.md:6]: nitpicky: but do we normally have a space here but not before the alerts?

File: windows-active-directory-mixin/README.md
Comment (StefanKurek) [inline windows-active-directory-mixin/README.md:40]: is this spacing incorrect (comparing it to the others)?

File: windows-observ-lib/targets.libsonnet
Comment (StefanKurek) [inline windows-observ-lib/targets.libsonnet:338]: should this be deleted?

File: windows-observ-lib/dashboards.libsonnet
Comment (StefanKurek) [inline windows-observ-lib/dashboards.libsonnet:31]: so what exactly is this and why is this panel special (why here for this one only?)

File: windows-observ-lib/panels.libsonnet
Comment (StefanKurek) [inline windows-observ-lib/panels.libsonnet:19]: just curious why all these lowercase changes?

File: windows-observ-lib/panels.libsonnet
Comment (StefanKurek) [inline windows-observ-lib/panels.libsonnet:453]: why are these commented out?

File: windows-observ-lib/panels.libsonnet
Comment (StefanKurek) [inline windows-observ-lib/panels.libsonnet:466]: any specific reason for these really detailed descriptions? i like them, but just want to check the reasoning

File: windows-observ-lib/dashboards.libsonnet
Comment (BominRahmani) [reply windows-observ-lib/dashboards.libsonnet:31]: This was when i was testing something out and forgot to get rid of it! Good catch!

File: windows-observ-lib/panels.libsonnet
Comment (BominRahmani) [reply windows-observ-lib/panels.libsonnet]: Not sure how this happened. It is unintended.

File: windows-observ-lib/panels.libsonnet
Comment (BominRahmani) [reply windows-observ-lib/panels.libsonnet]: This was my attempt at getting the alerts panel working previous to knowing how, since there is no alerts panel in the common lib. Had this commented out till i had more direction from Vitaly, Got it figured out now!

File: windows-active-directory-mixin/README.md
Comment (BominRahmani) [reply windows-active-directory-mixin/README.md:6]: I dont think it matters as the spacing changes depending on the mixin you are looking at, but i updated it to be more inline with the style you are thinking of i believe.

File: windows-observ-lib/panels.libsonnet
Comment (BominRahmani) [reply windows-observ-lib/panels.libsonnet:462]: I guess the reasoning is kinda circular, its more descriptive because i wanted them to be more descriptive?  On deeper thought its probably because these replication pending operations can include a variety of different things and I just wanted to include that into the description to more or less encapsulate that thought.

File: windows-active-directory-mixin/README.md
Comment (dpaasman00) [inline windows-active-directory-mixin/README.md:37]: Is this yaml section new? I don't remember it being in the README before

File: windows-active-directory-mixin/alerts/alerts.libsonnet
Comment (dpaasman00) [inline windows-active-directory-mixin/alerts/alerts.libsonnet:17]: I feel the second sentence being here defeats the purpose of this summary section. As far as I know and how I've treated it, this is just a quick blurb about the alert, so I feel the first sentence is enough for the summary. I think the second sentence would fit better in the description. Ultimately your call though if you've gotten review in the past that says otherwise. 

This comment extends to HighReplicationSyncRequestFailures and HighPasswordChanges

File: windows-active-directory-mixin/README.md
Comment (StefanKurek) [reply windows-active-directory-mixin/README.md:38]: @BominRahmani looks like you didn't respond to this one. i think the job name should be moved over to be in line with `windows_events`?

File: windows-observ-lib/panels.libsonnet
Comment (dpaasman00) [inline windows-observ-lib/panels.libsonnet:632]: For this panel and the intersite version, do they need the min, max, mean legend calcs?

File: windows-observ-lib/panels.libsonnet
Comment (StefanKurek) [inline windows-observ-lib/panels.libsonnet:43]: @BominRahmani looks like you forgot one?

File: windows-observ-lib/panels.libsonnet
Comment (dpaasman00) [reply windows-observ-lib/panels.libsonnet:632]: Same for inboundReplicationUpdates

**Review by dpaasman00 — COMMENTED** [review]
I think this looks good for the most part. I don't think I can give a proper review on the new library stuff, but what you have makes sense to me and the dashboard on the cloud seems fine to me. Just a few comments

File: windows-active-directory-mixin/README.md
Comment (BominRahmani) [reply windows-active-directory-mixin/README.md:37]: This yaml section is just to show what the log snippet would look like for the mixin. Its not new as far as I know and I have been asked to include it in the past.

File: windows-observ-lib/panels.libsonnet
Comment (BominRahmani) [reply windows-observ-lib/panels.libsonnet:632]: I added them in for now, I know previously I did not have them in, but I want to see what Grafana will say to this. The reason for the switch is I think having a calculated mean would be really helpful, and as far as i've seen whenever there is a calculated mean, there is always a min and max that follows it.

File: windows-active-directory-mixin/alerts/alerts.libsonnet
Comment (BominRahmani) [reply windows-active-directory-mixin/alerts/alerts.libsonnet:17]: Im not sure about this, I have always seen the description as being a flat out written form of where the thresholds have been crossed at what instances. The summary to me has always been why whatever alert has been triggered is an actual problem. I think I will keep it as is for now.

Comment (dpaasman00) [reply windows-active-directory-mixin/alerts/alerts.libsonnet:17]: That's fine, just something to keep in mind cus in the UI the summary doesn't have a lot of space, if you check out some of the pre-existing alerts on the cloud instance

File: windows-observ-lib/panels.libsonnet
Comment (dpaasman00) [reply windows-observ-lib/panels.libsonnet:632]: Fair enough to see what they say, but at that rate they may prefer a stat panel instead, based on feedback I got during spec review at one point when I had a similar situation ... just something to keep in mind

**Review by dpaasman00 — APPROVED** [review]

**Review by StefanKurek — APPROVED** [review]

**Comment by gaantunes** [comment]
Both dashboards are loading wrong default datasources, causing errors when we open them.

The logs dashboard does not show any data, tried 30 days back. Overall the overview dashboard looks good, would just like to see more data flowing if that is possible.

Screenshot from 2023-12-15 18-39-28
image

**Comment by BominRahmani** [comment]
> Both dashboards are loading wrong default datasources, causing errors when we open them.
> 
> The logs dashboard does not show any data, tried 30 days back. Overall the overview dashboard looks good, would just like to see more data flowing if that is possible.
> 
> Screenshot from 2023-12-15 18-39-28 image

Hey @gaantunes Not quite sure why the logs weren't showing up in that time range, i reset the VM and the logs are showing now so that should be good. As for the errors due to the default datasources, wouldn't this resolve itself on the integration itself similar to how the Log library resolves itself in the integration step? Maybe @v-zhuravlev you have some insights into this?

File: windows-observ-lib/panels.libsonnet
Comment (v-zhuravlev) [inline windows-observ-lib/panels.libsonnet:632]: Missing unit (bps?).

Tip: Since it is a network traffic of specific type, I think you can try to use `commonlib.panels.network.timeSeries.traffic` as a closer prototype here.

File: windows-observ-lib/panels.libsonnet
Comment (v-zhuravlev) [inline windows-observ-lib/panels.libsonnet:646]: missing unit.

File: windows-observ-lib/targets.libsonnet
Comment (v-zhuravlev) [inline windows-observ-lib/targets.libsonnet:361]: Perhaps we can be more generic here, so legend is transformed if something else, other than instance is used:

```suggestion
      + prometheusQuery.withLegendFormat(commonlib.utils.labelsToPanelLegend(this.config.instanceLabels)),
```

Please also add import to the top of the file:
`local commonlib = import 'common-lib/common/main.libsonnet';`

File: windows-observ-lib/targets.libsonnet
Comment (v-zhuravlev) [inline windows-observ-lib/targets.libsonnet:368]:
```suggestion
      + prometheusQuery.withLegendFormat('%s - {{ operation }}' % commonlib.utils.labelsToPanelLegend(this.config.instanceLabels)),
```
same as above

File: windows-observ-lib/targets.libsonnet
Comment (v-zhuravlev) [inline windows-observ-lib/targets.libsonnet:380]: You can add * 8 here, or change units from bps to Bps on the panel level instead.

```suggestion
        'rate(windows_ad_replication_data_intrasite_bytes_total{%(queriesSelector)s}[$__rate_interval]) * 8' % variables
```

**Review by v-zhuravlev — COMMENTED** [review]
Great work. Thanks for using modular libs approach.

Overall, I suggest to put not only panels/dashboards/targets into windows-observ-lib but also alerts group and config params.

That would leave folder `windows-active-directory-mixin` here as pure monitoring-mixin example, and nothing more.

When creating integration, we can import windows-observ-lib, and just set things like uid, filteringSelector, pick specific dashboards, etc... to generated required objects for AD integration.

File: windows-observ-lib/panels.libsonnet
Comment (v-zhuravlev) [inline windows-observ-lib/panels.libsonnet:456]: Could you try to use `queriesSelector` or at least this.config.instanceLabels, this.config.groupLabels to make sure that filter is changed accordingly?

One idea I have is to go to file variables.libsonnet and generate new field there called `queriesSelectorAdvancedSyntax` that will be just like `queriesSelector` but with ${<label>:regex} . Then we can use it here.

Comment (BominRahmani) [reply windows-observ-lib/panels.libsonnet:456]: Due to the way labelsToPromQLSelector works in commonlib,
image
doing it this way would cause the output to look like this
`job=~"integrations/windows_exporter",job:regex=~"$job:regex",instance:regex=~"$instance:regex"`
Since it reuses the label for the label and variable it will cause both parts to receive the regex modification that I add into the instance labels. Also the labelsToPromQLSelector function doesn't account for the `{}` encapsulation of the variable, eg: advanced grafana syntax should look like this: `'${job:regex}'`
Because of this i extended the commonlibs util function and added an advanced version of the labelsToPromQLSelector. This should take care of this, but I think it would be worth adding this function into commonlibs itself since alert panels are a regular occurence.

File: windows-active-directory-mixin/mixin.libsonnet
Comment (BominRahmani) [reply windows-active-directory-mixin/mixin.libsonnet]: Sorry I am not quite sure what you mean here, could you give an example?

**Comment by BominRahmani** [comment]
> Great work. Thanks for using modular libs approach.
> 
> Overall, I suggest to put not only panels/dashboards/targets into windows-observ-lib but also alerts group and config params.
> 
> That would leave folder `windows-active-directory-mixin` here as pure monitoring-mixin example, and nothing more.
> 
> When creating integration, we can import windows-observ-lib, and just set things like uid, filteringSelector, pick specific dashboards, etc... to generated required objects for AD integration.

I moved the alerts and configs into windows observ library, but I had some questions, I left the config option of enablingADDashboard on the mixin just to enable it since its defaulted to false in the library, im guessing this is ok? Also since the alerts are sitting within the observ library, wouldnt the windows exporter integration automatically pull all these alerts even though they may not have AD enabled?

File: windows-active-directory-mixin/mixin.libsonnet
Comment (v-zhuravlev) [reply windows-active-directory-mixin/mixin.libsonnet]: I think that those exrtra dashobard links are redundant - they are useful in three tier view, but here we only have two dashboards.

**Comment by BominRahmani** [comment]
> > > Great work. Thanks for using modular libs approach.
> > > Overall, I suggest to put not only panels/dashboards/targets into windows-observ-lib but also alerts group and config params.
> > > That would leave folder `windows-active-directory-mixin` here as pure monitoring-mixin example, and nothing more.
> > > When creating integration, we can import windows-observ-lib, and just set things like uid, filteringSelector, pick specific dashboards, etc... to generated required objects for AD integration.
> > 
> > 
> > I moved the alerts and configs into windows observ library, but I had some questions, I left the config option of enablingADDashboard on the mixin just to enable it since its defaulted to false in the library, im guessing this is ok? Also since the alerts are sitting within the observ library, wouldnt the windows exporter integration automatically pull all these alerts even though they may not have AD enabled?
> 
> yes, we can hide alerts there under AD feature flag as well.

Sounds good, I just added a check in the alerts section of windowsobservlib which checks whether that feature flag is enabled. I think this is what you had in mind yes? @v-zhuravlev

File: windows-mixin/config.libsonnet
Comment (v-zhuravlev) [inline windows-mixin/config.libsonnet:22]:
```suggestion
    enableADDashboard: false,
    // set to false to disable logs dashboard and logs annotations
```

**Review by v-zhuravlev — APPROVED** [review]

**Pull request merged**
